File: src/lib/vectorstore.ts
import { env } from '@/env'
import { OpenAIEmbeddings } from '@langchain/openai'
import { PineconeStore } from '@langchain/pinecone'
import { Pinecone } from '@pinecone-database/pinecone'

// initialize pinecone client
export const pinecone = new Pinecone({
  apiKey: env.PINECONE_API_KEY,
})

// create embeddings client
export const embeddings = new OpenAIEmbeddings({
  openAIApiKey: env.HEURIST_AUTH_KEY,
  modelName: 'BAAI/bge-large-en-v1.5',
  configuration: {
    baseURL: env.HEURIST_GATEWAY_URL || 'https://llm-gateway.heurist.xyz',
  },
})

// get or initialize the vector store
export async function getVectorStore() {
  return PineconeStore.fromExistingIndex(embeddings, {
    pineconeIndex: pinecone.index(env.PINECONE_INDEX),
  })
}

File: backend/app/services/vectorstore.py
import os
from typing import List

from langchain_pinecone import PineconeVectorStore
from openai import OpenAI
from pinecone import Pinecone

from ..utils.logger import logger


class CustomEmbeddings:
    def __init__(self):
        self.client = OpenAI(
            api_key=os.getenv("HEURIST_AUTH_KEY"),
            base_url=os.getenv(
                "HEURIST_GATEWAY_URL", "https://llm-gateway.heurist.xyz"
            ),
        )

    def embed_documents(self, texts: List[str]) -> List[List[float]]:
        logger.debug(f"Embedding {len(texts)} documents")
        response = self.client.embeddings.create(
            model="BAAI/bge-large-en-v1.5", input=texts, encoding_format="float"
        )
        return [data.embedding for data in response.data]

    def embed_query(self, text: str) -> List[float]:
        logger.debug("Embedding query")
        response = self.client.embeddings.create(
            model="BAAI/bge-large-en-v1.5", input=text, encoding_format="float"
        )
        return response.data[0].embedding


logger.info("Initializing Pinecone client")
pinecone = Pinecone(api_key=os.getenv("PINECONE_API_KEY"))

logger.info("Creating embeddings client")
embeddings = CustomEmbeddings()

logger.info("Initializing vector store")
try:
    vector_store = PineconeVectorStore.from_existing_index(
        index_name=os.getenv("PINECONE_INDEX"),
        embedding=embeddings,
    )
    logger.info("Vector store initialized successfully")
except Exception as e:
    logger.error(f"Error initializing vector store: {e}", exc_info=True)
    raise


def get_vector_store():
    """Get vector store instance"""
    return vector_store
